Drain nodes in batches

alkar opened the pull request.

alkar committed:
Add parse_opts and checkdeps functions

Tidy up these two distinct initialisation steps in their own separate
functions.

Signed-off-by: Dimitrios Karagiannis <dhkarag@gmail.com>

alkar committed:
Fix various issues detected by shellcheck

Signed-off-by: Dimitrios Karagiannis <dhkarag@gmail.com>

alkar committed:
Extract upscale_asg and downscale_asg funcs

Extract the logic in separate, smaller functions for clarity.

Signed-off-by: Dimitrios Karagiannis <dhkarag@gmail.com>

alkar committed:
Extract drain_nodes function

The same operation is used by both normal and resume modes. Also, use
$retire_time in both modes and $resume only for determining the mode,
which should make it a bit more clear.

Signed-off-by: Dimitrios Karagiannis <dhkarag@gmail.com>

alkar committed:
Merge cycle_nodes in update

There is no real reason to keep these two functions separate.

Additionally, optimise the ASG discovery steps:
- Call describe-auto-scaling-groups once
- Set asg_count to the MinSize property. This does not change and can be
used during the resume operation.

Signed-off-by: Dimitrios Karagiannis <dhkarag@gmail.com>

alkar committed:
Revisit resume operation

Resume should also call downscale_asg when it's finished to avoid the
need for manual steps. We can simply call update and if the resume flag
is set then we can skip the unncessary steps.

Signed-off-by: Dimitrios Karagiannis <dhkarag@gmail.com>

alkar committed:
Retry ASG discovery on failure

Occasionally when resuming shortly after a node has been terminated, we cannot
find the ASG name since the first node in the list is still part of the cluster
but not part of the ASG. In that case, rather than failing hard, we can
retry until it succeeds.

Signed-off-by: Dimitrios Karagiannis <dhkarag@gmail.com>

alkar committed:
Do not suspend the ReplaceUnhealthy process

The AWS documentation mentions that when the Launch process is
suspended, "ReplaceUnhealthy continues to terminate unhealthy instances,
but does not launch replacements". We actually terminate instances
manually, through this script but the ReplaceUnhealthy process allows
the ASG to acknowledge these terminated instances. It actually does a
no-op termination itself before removing them from the ASG pool.

With this change, we should now resize the ASG _after_ the instance
count has reached the desired number which should result in less
turbulence after resuming the previously suspended processes.

Signed-off-by: Dimitrios Karagiannis <dhkarag@gmail.com>

alkar committed:
Drain nodes in batches to avoid large scale outs

Doubling the size of the ASG is only practical up to a limit. That, in
addition with the likelihood of bumping onto AWS capacity issues
dictates a different solution.

With this change, we aim to drain nodes in batches (of three, by
default). This means that we only need to scale out by a number of nodes
equal to the "batch size" and allow the ASG to replace terminated
nodes with new ones as we go.

Signed-off-by: Dimitrios Karagiannis <dhkarag@gmail.com>

alkar committed:
Consider AZ distribution when draining nodes

There have been cases where draining a number of nodes from a single AZ
in quick succession has produced a lopsided ASG. Since we don't control
(or necessarily know) how/where the ASG decides to launch new instances,
we can aim to drain them evenly, in order to help it make more
intelligent decisions.

Signed-off-by: Dimitrios Karagiannis <dhkarag@gmail.com>

george-angel commented on kube-aws-updater:
```diff
@@ -18,97 +18,105 @@ set -o errexit
         must also be set
   -t    Node drain timeout. How long to wait for the node to drain before
         shutting it down (in seconds, default ${timeout}s)
+  -n    Node batch size. How many nodes to add to the ASG. Also, how many nodes
```
This should specify what the default value is

george-angel commented on kube-aws-updater:
```diff
@@ -246,45 +237,138 @@ cycle_nodes() {
+    ic=$(retry aws "${aws_opts[@]}" --output=json autoscaling describe-auto-scaling-groups \
       --auto-scaling-group-names "${asg_name}" |\
-      jq -r -e '."AutoScalingGroups"[0]."Instances" | length')
+      jq -r -e '.AutoScalingGroups[0].Instances | length')
```
I don't mind general format changes, but why this one?

alkar replied:
For consistency with the rest of `jq` queries and because I find it very odd and confusing that it's quoted when there's no need for quotes.

alkar committed:
node_batch_size: check for greater than zero value

Also, mention the default in usage output.

Signed-off-by: Dimitrios Karagiannis <dhkarag@gmail.com>

george-angel commented on kube-aws-updater:
```diff
@@ -246,45 +241,138 @@ cycle_nodes() {
+        break
+    fi
+  done
+  log "asg_name: '${asg_name}' / asg_count: '${asg_count}'"
```
nitpick: this is an odd format.

Although not mentioned, I tried to keep hclog fmt: `message text: key1=value1 key2=value2` or `message text`

alkar committed:
Explain draining in detail, with examples

Signed-off-by: Dimitrios Karagiannis <dhkarag@gmail.com>

alkar committed:
Convert to hclog format

Signed-off-by: Dimitrios Karagiannis <dhkarag@gmail.com>

alkar committed:
Use the same node naming scheme in comments

Signed-off-by: Dimitrios Karagiannis <dhkarag@gmail.com>

alkar committed:
Allow node balance to be off by one

This is to cater for master node groups whose size is not a multiple of
three (eg.: 5, 7, 11).

Signed-off-by: Dimitrios Karagiannis <dhkarag@gmail.com>

alkar committed:
Introduce grace_delta for the node balance check

Signed-off-by: Dimitrios Karagiannis <dhkarag@gmail.com>
